lib/plug/conn.ex

```diff
@@ -1328,50 +1340,148 @@ defmodule Plug.Conn do
+  `Base.decode64(encoded, padding: false)` when reading the cookie.
+  It is important for padding to be disabled since `=` is not a valid
+  character in cookie values.
+
+  ## Signing and encrypting cookies
+
+  This function allows you to automatically sign and encrypt cookies.
+  When signing or encryption is enabled, then any Elixir value can be
+  stored in the cookie (except anonymous functions for security reasons).
+  Once a value is signed or encrypted, you must also call `fetch_cookie/2`
+  with the cookies that are either signed or encrypted.
+
+  To sign, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, sign: true)
+
+  and then:
+
+      fetch_cookie(conn, signed: ~w(my-cookie))
+
+  To encrypt, you would do:
+
+      put_resp_cookie(conn, "my-cookie", %{user_id: user.id}, encrypt: true)
+
+  and then:
+
+      fetch_cookie(conn, encrypted: ~w(my-cookie))
+
+  By default a signed or encrypted cookie is only valid for a day, unless
+  a `:max_age` is specified.
 
   ## Options
 
     * `:domain` - the domain the cookie applies to
     * `:max_age` - the cookie max-age, in seconds. Providing a value for this
-      option will set both the _max-age_ and _expires_ cookie attributes
+      option will set both the _max-age_ and _expires_ cookie attributes.
     * `:path` - the path the cookie applies to
     * `:http_only` - when `false`, the cookie is accessible beyond HTTP
     * `:secure` - if the cookie must be sent only over https. Defaults
       to true when the connection is HTTPS
     * `:extra` - string to append to cookie. Use this to take advantage of
       non-standard cookie attributes.
+    * `:signed` - when true, signs the
 
   """
   @spec put_resp_cookie(t, binary, binary, Keyword.t()) :: t
   def put_resp_cookie(%Conn{} = conn, key, value, opts \\ [])
-      when is_binary(key) and is_binary(value) and is_list(opts) do
+      when is_binary(key) and is_list(opts) do
     %{resp_cookies: resp_cookies, scheme: scheme} = conn
-    cookie = [{:value, value} | opts] |> :maps.from_list() |> maybe_secure_cookie(scheme)
+    {to_send_value, opts} = maybe_sign_or_encrypt_cookie(conn, key, value, opts)
+    cookie = [{:value, to_send_value} | opts] |> Map.new() |> maybe_secure_cookie(scheme)
     resp_cookies = Map.put(resp_cookies, key, cookie)
     update_cookies(%{conn | resp_cookies: resp_cookies}, &Map.put(&1, key, value))
   end
 
+  defp maybe_sign_or_encrypt_cookie(conn, key, value, opts) do
+    {sign?, opts} = Keyword.pop(opts, :sign, false)
+    {encrypt?, opts} = Keyword.pop(opts, :encrypt, false)
+
+    case {sign?, encrypt?} do
+      {true, true} ->
+        raise ArgumentError, "only :sign or :encrypt may be given at the same time"
+
+      {true, false} ->
+        {Plug.Crypto.sign(conn.secret_key_base, key, value, signed_at(opts)), opts}
+
+      {false, true} ->
+        {Plug.Crypto.encrypt(conn.secret_key_base, key, key, value, signed_at(opts)), opts}
```

Comment:
@GriffinMB @voltone  - I am using the cookie key as the encryption secret and signing salt. Is this a good or bad idea?

Reply:
Using the same salt for encryption and signing doesn't seem great. You could perhaps derive two different values from the cookie name, using prefixes. Again, I'll think about it...

Reply:
This will generate the same secret for the encryption "secret" and "sign_secret". The sign secret is ultimately used as GCM-modes AAD. I'm not aware of any attacks on this, but it is definitely not the intended use, so may break the security proofs. 

I would use some other value here; though, more broadly, I don't think the "sign_secret" is really necessary for encryption at all so we could remove it.

Reply:
It will be great if we could remove it, as it would simplify the API, but I have no idea about the implication of removing it in Plug.Crypto. For now I will add a different suffix to each.

Reply:
The RFC [actually says](https://tools.ietf.org/html/rfc7518#section-4.7): "The Additional Authenticated Data value used is the empty octet string."  So setting encryption_salt to `key` and signing_salt to an empty string should also be ok.

Reply:
Removing the sign_secret in Plug.Crypto wouldn't have any security effect, but would be a breaking API change. And, yes, in this instance you can make the "sign_secret" anything you like - an empty string, a version number, etc

Reply:
Well, actually you need it for namespacing. It just doesn’t need to be secret or random. So just an internal API change.

I’ll confirm and make a PR to Plug.Crypto later today!

Reply:
You **could** leave it in for namespacing/versioning/whatever, which is what the AAD is intended to be used for, which wouldn't necessarily be a breaking change. I'll take a look later today and make a PR on Plug.Crypto.

Reply:
Unfortunately if we stop using it for its current purpose it will be a breaking change, so my proposal is to remove it from `Plug.Crypto.encrypt/decrypt` only but still expose it on lower level stuff.

Reply:
@GriffinMB I hid the salt in the higher level API while keeping the old one around here: https://github.com/elixir-plug/plug_crypto/commit/be3ef080e7e55e90f1cb43bf41352f89a0c08a8a

WDYT?